server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/DomainsRoutes.java

```diff
@@ -212,16 +283,58 @@ private void addDomain(Domain domain) throws DomainListException {
     }
 
     private Response exists(Request request, Response response) throws DomainListException {
-        Domain domain = checkValidDomain(request);
+        Domain domain = checkValidDomain(request.params(DOMAIN_NAME));
+
         if (!domainList.containsDomain(domain)) {
-            throw ErrorResponder.builder()
-                .statusCode(HttpStatus.NOT_FOUND_404)
-                .type(ErrorType.INVALID_ARGUMENT)
-                .message("The domain list does not contain: " + domain.name())
-                .haltError();
+            throw domainNotFound(domain);
         } else {
             response.status(HttpStatus.NO_CONTENT_204);
             return response;
         }
     }
+
+    private ImmutableSet<DomainAliasResponse> listDomainAliases(Request request, Response response) throws DomainListException, RecipientRewriteTableException {
+        Domain domain = checkValidDomain(request.params(DOMAIN_NAME));
+
+        if (!domainList.containsDomain(domain)) {
+            throw domainNotFound(domain);
+        } else {
+            return recipientRewriteTable.listSources(Mapping.domain(domain))
+                .sorted(Comparator.comparing(MappingSource::asMailAddressString))
+                .map(DomainAliasResponse::new)
+                .collect(Guavate.toImmutableSet());
+        }
+    }
+
+    private String addDomainAlias(Request request, Response response) throws DomainListException, RecipientRewriteTableException {
+        return performOperationOnAlias(request, response, recipientRewriteTable::addMapping);
+    }
+
+    private String removeDomainAlias(Request request, Response response) throws DomainListException, RecipientRewriteTableException {
+        return performOperationOnAlias(request, response, recipientRewriteTable::removeMapping);
+    }
+
+    private String performOperationOnAlias(Request request, Response response, MappingOperation operation) throws DomainListException, RecipientRewriteTableException {
+        Domain sourceDomain = checkValidDomain(request.params(SOURCE_DOMAIN));
+        Domain destinationDomain = checkValidDomain(request.params(DESTINATION_DOMAIN));
+
+        if (!domainList.containsDomain(sourceDomain)) {
+            throw domainNotFound(sourceDomain);
+        }
+        if (!domainList.containsDomain(destinationDomain)) {
```

Comment:
I think we should only check the sourceDomain existence, a destination domain not by managed by James is still valid. For example I send and email to source address abc@james.org and the mail goes to the destination abc@gmail.com (still valid)

Reply:
True - let's relax this...